ott/core/initializers.py

```diff
@@ -0,0 +1,260 @@
+    init_dual_a = jnp.zeros_like(a) if lse_mode else jnp.ones_like(a)
+    return init_dual_a
+
+  def init_dual_b(
+      self, ot_problem: linear_problems.LinearProblem, lse_mode: bool
+  ) -> jnp.ndarray:
+    """Initialization for Sinkhorn potential/ scaling g_v.
+
+    Args:
+      ot_problem: OT problem between discrete distributions of size n and m.
+      lse_mode: Return potential if true, scaling if false.
+
+    Returns:
+      potential/ scaling, array of size m
+    """
+    b = ot_problem.b
+    init_dual_b = jnp.zeros_like(b) if lse_mode else jnp.ones_like(b)
+    return init_dual_b
+
+
+class GaussianInitializer(DefaultInitializer):
+  """GaussianInitializer.
+
+  From :cite:`thornton2022rethinking:22`.
+  Compute Gaussian approximations of each pointcloud, then compute closed from
+  Kantorovich potential betwen Gaussian approximations using Brenier's theorem
+  (adapt convex/ Brenier potential to Kantorovich). Use this Gaussian potential to
+  initialize Sinkhorn potentials/ scalings.
+
+  """
+
+  def __init__(self):
+    super().__init__()
+
+  def init_dual_a(
+      self,
+      ot_problem: linear_problems.LinearProblem,
+      lse_mode: bool,
+  ) -> jnp.ndarray:
+    """Gaussian init function.
+
+    Args:
+      ot_problem: OT problem description with geometry and weights.
+      lse_mode: Return potential if true, scaling if false.
+
+    Returns:
+      potential/ scaling f_u, array of size n.
+    """
+    # import Gaussian here due to circular imports
+    from ott.tools.gaussian_mixture import gaussian
+
+    assert isinstance(
+        ot_problem.geom, pointcloud.PointCloud
+    ), "Gaussian initializer valid only for PointCloud geom"
+
+    x, y = ot_problem.geom.x, ot_problem.geom.y
+    a, b = ot_problem.a, ot_problem.b
+
+    gaussian_a = gaussian.Gaussian.from_samples(x, weights=a)
+    gaussian_b = gaussian.Gaussian.from_samples(y, weights=b)
+    # Brenier potential for cost ||x-y||^2/2, multiply by two for ||x-y||^2
+    f_potential = 2 * gaussian_a.f_potential(dest=gaussian_b, points=x)
+    f_potential = f_potential - jnp.mean(f_potential)
+    f_u = f_potential if lse_mode else ot_problem.scaling_from_potential(
+        f_potential
+    )
+    return f_u
+
+
+class SortingInitializer(DefaultInitializer):
+  """Sorting Init class.
+
+  DualSort algorithm from :cite:`thornton2022rethinking:22`, solve
+  non-regularized OT problem via sorting, then compute potential through
+  iterated minimum on C-transform and use this potential to initialize
+  regularized potential
+
+  Args:
+    vectorized_update: Use vectorized inner loop if true.
+    tolerance: DualSort convergence threshold.
+    max_iter: Max DualSort steps.
+  """
+
+  def __init__(
+      self,
+      vectorized_update: bool = True,
+      tolerance: float = 1e-2,
+      max_iter: int = 100
+  ):
+    super().__init__()
+    self.tolerance = tolerance
+    self.max_iter = max_iter
+    self.update_fn = lambda f, mod_cost: jax.lax.cond(
```

Comment:
Why not just `self.update_fn = _vectorized_update if vectorized_update else _coordinate_update`? Should be traceable by JAX, since `vectorized_update` is a Python constant.

Reply:
There is some difference in the XLA, just glancing over it, the condition is checked every iteration during runtime, which will a bit slower.

`jax.lax.cond`

<details>

```python
{ lambda ; a:f32[128,3] b:f32[128,3]. let
    c:f32[128,128] = dot_general[
      dimension_numbers=(((1,), (1,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] a b
    d:f32[128,128] = mul -2.0 c
    e:f32[128,3] = integer_pow[y=2] a
    f:f32[128] = reduce_sum[axes=(1,)] e
    g:f32[128,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(128, 1)] f
    h:f32[128,3] = integer_pow[y=2] b
    i:f32[128] = reduce_sum[axes=(1,)] h
    j:f32[1,128] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 128)] i
    k:f32[128,128] = add g j
    l:f32[128,128] = add d k
    m:f32[128,128] = pow l 1.0
    n:f32[128,128] = mul m 1.0
    o:f32[128] = xla_call[
      call_jaxpr={ lambda ; p:f32[128,128]. let
          q:f32[128] = xla_call[
            call_jaxpr={ lambda ; r:f32[128,128]. let
                s:i32[128] = iota[dimension=0 dtype=int32 shape=(128,)] 
                t:i32[128] = iota[dimension=0 dtype=int32 shape=(128,)] 
                u:bool[128] = lt s 0
                v:i32[128] = add s 128
                w:i32[128] = select_n u s v
                x:bool[128] = lt t 0
                y:i32[128] = add t 128
                z:i32[128] = select_n x t y
                ba:i32[128,1] = broadcast_in_dim[
                  broadcast_dimensions=(0,)
                  shape=(128, 1)
                ] w
                bb:i32[128,1] = broadcast_in_dim[
                  broadcast_dimensions=(0,)
                  shape=(128, 1)
                ] z
                bc:i32[128,2] = concatenate[dimension=1] ba bb
                bd:f32[128] = gather[
                  dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0, 1), start_index_map=(0, 1))
                  fill_value=None
                  indices_are_sorted=False
                  mode=GatherScatterMode.PROMISE_IN_BOUNDS
                  slice_sizes=(1, 1)
                  unique_indices=False
                ] r bc
              in (bd,) }
            name=diagonal
          ] p
        in (q,) }
      name=_diag
    ] n
    be:f32[1,128] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 128)] o
    bf:f32[128,128] = sub n be
    bg:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 0.0
    bh:f32[128] _:f32[] _:i32[] = while[
      body_jaxpr={ lambda ; bi:f32[128,128] bj:f32[128] bk:f32[] bl:i32[]. let
          bm:f32[128] = cond[
            branches=(
              { lambda ; bn:f32[128] bo:f32[128,128]. let
                  _:i32[] bp:f32[128] = scan[
                    jaxpr={ lambda ; bq:f32[128,128] br:i32[] bs:f32[128]. let
                        bt:i32[] = add br 1
                        bu:bool[] = lt br 0
                        bv:i32[] = add br 128
                        bw:i32[] = select_n bu br bv
                        bx:i32[] = convert_element_type[
                          new_dtype=int32
                          weak_type=False
                        ] bw
                        by:i32[1] = broadcast_in_dim[
                          broadcast_dimensions=()
                          shape=(1,)
                        ] bx
                        bz:f32[128] = gather[
                          dimension_numbers=GatherDimensionNumbers(offset_dims=(0,), collapsed_slice_dims=(0,), start_index_map=(0,))
                          fill_value=None
                          indices_are_sorted=True
                          mode=GatherScatterMode.PROMISE_IN_BOUNDS
                          slice_sizes=(1, 128)
                          unique_indices=True
                        ] bq by
                        ca:f32[128] = add bz bs
                        cb:f32[] = reduce_min[axes=(0,)] ca
                        cc:bool[] = lt br 0
                        cd:i32[] = add br 128
                        ce:i32[] = select_n cc br cd
                        cf:i32[] = convert_element_type[
                          new_dtype=int32
                          weak_type=False
                        ] ce
                        cg:i32[1] = broadcast_in_dim[
                          broadcast_dimensions=()
                          shape=(1,)
                        ] cf
                        ch:f32[128] = scatter[
                          dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
                          indices_are_sorted=True
                          mode=GatherScatterMode.FILL_OR_DROP
                          unique_indices=True
                          update_consts=()
                          update_jaxpr={ lambda ; ci:f32[] cj:f32[]. let
                              
                            in (cj,) }
                        ] bs cg cb
                      in (bt, ch) }
                    length=128
                    linear=(False, False, False)
                    num_carry=2
                    num_consts=1
                    reverse=False
                    unroll=1
                  ] bo 0 bn
                in (bp,) }
              { lambda ; ck:f32[128] cl:f32[128,128]. let
                  cm:f32[1,128] = broadcast_in_dim[
                    broadcast_dimensions=(1,)
                    shape=(1, 128)
                  ] ck
                  cn:f32[128,128] = add cl cm
                  co:f32[128] = reduce_min[axes=(1,)] cn
                in (co,) }
            )
            linear=(False, False)
          ] 1 bj bi
          cp:f32[128] = sub bm bj
          cq:f32[128] = integer_pow[y=2] cp
          cr:f32[] = reduce_sum[axes=(0,)] cq
          cs:i32[] = add bl 1
        in (bm, cr, cs) }
      body_nconsts=1
      cond_jaxpr={ lambda ; ct:f32[128] cu:f32[] cv:i32[]. let
          cw:bool[] = gt cu 0.009999999776482582
          cx:bool[] = lt cv 100
          cy:bool[] = convert_element_type[new_dtype=bool weak_type=False] cw
          cz:bool[] = convert_element_type[new_dtype=bool weak_type=False] cx
          da:bool[] = and cy cz
        in (da,) }
      cond_nconsts=0
    ] bf bg 1.01 0
    db:f32[] = reduce_sum[axes=(0,)] bh
    dc:f32[] = div db 128.0
    dd:f32[128] = sub bh dc
    de:bool[128] = xla_call[
      call_jaxpr={ lambda ; df:f32[128]. let
          dg:bool[128] = is_finite df
        in (dg,) }
      name=isfinite
    ] dd
    dh:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 1.0
    di:f32[] = reduce_sum[axes=(0,)] dh
    dj:f32[128] = div dh di
    dk:f32[128,3] = integer_pow[y=2] a
    dl:f32[128] = reduce_sum[axes=(1,)] dk
    dm:f32[128,3] = integer_pow[y=2] b
    dn:f32[128] = reduce_sum[axes=(1,)] dm
    do:f32[] = dot_general[
      dimension_numbers=(((0,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] dl dj
    dp:f32[1,1] = reshape[dimensions=None new_sizes=(1, 1)] do
    dq:f32[128,1] = reshape[dimensions=None new_sizes=(128, 1)] dn
    dr:f32[] = reduce_sum[axes=(0,)] dj
    ds:f32[1,1] = reshape[dimensions=None new_sizes=(1, 1)] dr
    dt:f32[128,1] = mul dq ds
    du:f32[128,1] = add dp dt
    dv:f32[3,128] = transpose[permutation=(1, 0)] a
    dw:f32[3] = dot_general[
      dimension_numbers=(((1,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] dv dj
    dx:f32[128] = dot_general[
      dimension_numbers=(((1,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] b dw
    dy:f32[128] = mul -2.0 dx
    dz:f32[128,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(128, 1)] dy
    ea:f32[128,1] = add du dz
    eb:f32[128,1] = mul 1.0 ea
    ec:f32[128] = squeeze[dimensions=(1,)] eb
    ed:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 1.0
    ee:f32[] = reduce_sum[axes=(0,)] ed
    ef:f32[128] = div ed ee
    eg:f32[128] = mul ec ef
    eh:f32[] = reduce_sum[axes=(0,)] eg
    ei:f32[] = stop_gradient eh
    ej:f32[] = xla_call[
      call_jaxpr={ lambda ; ek:bool[] el:f32[] em:f32[]. let
          en:f32[] = convert_element_type[new_dtype=float32 weak_type=False] em
          eo:f32[] = select_n ek en el
        in (eo,) }
      name=_where
    ] True ei 1.0
    ep:f32[] = mul 0.05000000074505806 ej
    eq:f32[128] = div dd ep
    er:f32[128] = xla_call[
      call_jaxpr={ lambda ; es:bool[128] et:f32[128] eu:f32[]. let
          ev:f32[] = convert_element_type[new_dtype=float32 weak_type=False] eu
          ew:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] ev
          ex:f32[128] = select_n es ew et
        in (ex,) }
      name=_where
    ] de eq 0.0
    ey:f32[128] = exp er
    ez:f32[128] = xla_call[
      call_jaxpr={ lambda ; fa:bool[128] fb:f32[128] fc:f32[]. let
          fd:f32[] = convert_element_type[new_dtype=float32 weak_type=False] fc
          fe:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] fd
          ff:f32[128] = select_n fa fe fb
        in (ff,) }
      name=_where
    ] de ey 0.0
  in (ez,) }
```

</details>

`Python if`

<details>

```python
{ lambda ; a:f32[128,3] b:f32[128,3]. let
    c:f32[128,128] = dot_general[
      dimension_numbers=(((1,), (1,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] a b
    d:f32[128,128] = mul -2.0 c
    e:f32[128,3] = integer_pow[y=2] a
    f:f32[128] = reduce_sum[axes=(1,)] e
    g:f32[128,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(128, 1)] f
    h:f32[128,3] = integer_pow[y=2] b
    i:f32[128] = reduce_sum[axes=(1,)] h
    j:f32[1,128] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 128)] i
    k:f32[128,128] = add g j
    l:f32[128,128] = add d k
    m:f32[128,128] = pow l 1.0
    n:f32[128,128] = mul m 1.0
    o:f32[128] = xla_call[
      call_jaxpr={ lambda ; p:f32[128,128]. let
          q:f32[128] = xla_call[
            call_jaxpr={ lambda ; r:f32[128,128]. let
                s:i32[128] = iota[dimension=0 dtype=int32 shape=(128,)] 
                t:i32[128] = iota[dimension=0 dtype=int32 shape=(128,)] 
                u:bool[128] = lt s 0
                v:i32[128] = add s 128
                w:i32[128] = select_n u s v
                x:bool[128] = lt t 0
                y:i32[128] = add t 128
                z:i32[128] = select_n x t y
                ba:i32[128,1] = broadcast_in_dim[
                  broadcast_dimensions=(0,)
                  shape=(128, 1)
                ] w
                bb:i32[128,1] = broadcast_in_dim[
                  broadcast_dimensions=(0,)
                  shape=(128, 1)
                ] z
                bc:i32[128,2] = concatenate[dimension=1] ba bb
                bd:f32[128] = gather[
                  dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0, 1), start_index_map=(0, 1))
                  fill_value=None
                  indices_are_sorted=False
                  mode=GatherScatterMode.PROMISE_IN_BOUNDS
                  slice_sizes=(1, 1)
                  unique_indices=False
                ] r bc
              in (bd,) }
            name=diagonal
          ] p
        in (q,) }
      name=_diag
    ] n
    be:f32[1,128] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 128)] o
    bf:f32[128,128] = sub n be
    bg:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 0.0
    bh:f32[128] _:f32[] _:i32[] = while[
      body_jaxpr={ lambda ; bi:f32[128,128] bj:f32[128] bk:f32[] bl:i32[]. let
          bm:f32[1,128] = broadcast_in_dim[
            broadcast_dimensions=(1,)
            shape=(1, 128)
          ] bj
          bn:f32[128,128] = add bi bm
          bo:f32[128] = reduce_min[axes=(1,)] bn
          bp:f32[128] = sub bo bj
          bq:f32[128] = integer_pow[y=2] bp
          br:f32[] = reduce_sum[axes=(0,)] bq
          bs:i32[] = add bl 1
        in (bo, br, bs) }
      body_nconsts=1
      cond_jaxpr={ lambda ; bt:f32[128] bu:f32[] bv:i32[]. let
          bw:bool[] = gt bu 0.009999999776482582
          bx:bool[] = lt bv 100
          by:bool[] = convert_element_type[new_dtype=bool weak_type=False] bw
          bz:bool[] = convert_element_type[new_dtype=bool weak_type=False] bx
          ca:bool[] = and by bz
        in (ca,) }
      cond_nconsts=0
    ] bf bg 1.01 0
    cb:f32[] = reduce_sum[axes=(0,)] bh
    cc:f32[] = div cb 128.0
    cd:f32[128] = sub bh cc
    ce:bool[128] = xla_call[
      call_jaxpr={ lambda ; cf:f32[128]. let
          cg:bool[128] = is_finite cf
        in (cg,) }
      name=isfinite
    ] cd
    ch:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 1.0
    ci:f32[] = reduce_sum[axes=(0,)] ch
    cj:f32[128] = div ch ci
    ck:f32[128,3] = integer_pow[y=2] a
    cl:f32[128] = reduce_sum[axes=(1,)] ck
    cm:f32[128,3] = integer_pow[y=2] b
    cn:f32[128] = reduce_sum[axes=(1,)] cm
    co:f32[] = dot_general[
      dimension_numbers=(((0,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] cl cj
    cp:f32[1,1] = reshape[dimensions=None new_sizes=(1, 1)] co
    cq:f32[128,1] = reshape[dimensions=None new_sizes=(128, 1)] cn
    cr:f32[] = reduce_sum[axes=(0,)] cj
    cs:f32[1,1] = reshape[dimensions=None new_sizes=(1, 1)] cr
    ct:f32[128,1] = mul cq cs
    cu:f32[128,1] = add cp ct
    cv:f32[3,128] = transpose[permutation=(1, 0)] a
    cw:f32[3] = dot_general[
      dimension_numbers=(((1,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] cv cj
    cx:f32[128] = dot_general[
      dimension_numbers=(((1,), (0,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] b cw
    cy:f32[128] = mul -2.0 cx
    cz:f32[128,1] = broadcast_in_dim[broadcast_dimensions=(0,) shape=(128, 1)] cy
    da:f32[128,1] = add cu cz
    db:f32[128,1] = mul 1.0 da
    dc:f32[128] = squeeze[dimensions=(1,)] db
    dd:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] 1.0
    de:f32[] = reduce_sum[axes=(0,)] dd
    df:f32[128] = div dd de
    dg:f32[128] = mul dc df
    dh:f32[] = reduce_sum[axes=(0,)] dg
    di:f32[] = stop_gradient dh
    dj:f32[] = xla_call[
      call_jaxpr={ lambda ; dk:bool[] dl:f32[] dm:f32[]. let
          dn:f32[] = convert_element_type[new_dtype=float32 weak_type=False] dm
          do:f32[] = select_n dk dn dl
        in (do,) }
      name=_where
    ] True di 1.0
    dp:f32[] = mul 0.05000000074505806 dj
    dq:f32[128] = div cd dp
    dr:f32[128] = xla_call[
      call_jaxpr={ lambda ; ds:bool[128] dt:f32[128] du:f32[]. let
          dv:f32[] = convert_element_type[new_dtype=float32 weak_type=False] du
          dw:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] dv
          dx:f32[128] = select_n ds dw dt
        in (dx,) }
      name=_where
    ] ce dq 0.0
    dy:f32[128] = exp dr
    dz:f32[128] = xla_call[
      call_jaxpr={ lambda ; ea:bool[128] eb:f32[128] ec:f32[]. let
          ed:f32[] = convert_element_type[new_dtype=float32 weak_type=False] ec
          ee:f32[128] = broadcast_in_dim[broadcast_dimensions=() shape=(128,)] ed
          ef:f32[128] = select_n ea ee eb
        in (ef,) }
      name=_where
    ] ce dy 0.0
  in (dz,) }
```

</details>

Reply:
Locally this doesn't give me any problems, `test_sorting_init` fails because the `vector_min` is not marked as a `static` argument.